autoprecompiles/src/memory_optimizer.rs

```diff
@@ -0,0 +1,323 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt;
+use std::fmt::Display;
+use std::hash::Hash;
+
+use itertools::Itertools;
+use powdr_ast::analyzed::{
+    algebraic_expression_conversion, AlgebraicBinaryOperator, AlgebraicExpression,
+    AlgebraicReference, Challenge,
+};
+use powdr_constraint_solver::quadratic_symbolic_expression::RangeConstraintProvider;
+use powdr_constraint_solver::range_constraint::RangeConstraint;
+use powdr_constraint_solver::{
+    quadratic_symbolic_expression::QuadraticSymbolicExpression,
+    symbolic_expression::SymbolicExpression,
+};
+use powdr_number::{FieldElement, LargeInt};
+
+use crate::{MemoryBusInteraction, MemoryType, SymbolicConstraint, SymbolicMachine};
+
+pub fn optimize_memory<T: FieldElement>(mut machine: SymbolicMachine<T>) -> SymbolicMachine<T> {
+    let memory_bus_interactions = machine
+        .bus_interactions
+        .iter()
+        .filter_map(|bus| {
+            let mem_int = MemoryBusInteraction::try_from(bus.clone()).ok()?;
+            if matches!(mem_int.ty, MemoryType::Memory) {
+                Some(mem_int)
+            } else {
+                None
+            }
+        })
+        .collect_vec();
+
+    // Transform the constraints such that some quadratic constraints are turned into
+    // affine when we introduce a new boolean variable ("zero check variable").
+    let mut zero_check_transformer = ZeroCheckTransformer::default();
+    let constraints = machine
+        .constraints
+        .iter()
+        .map(|constr| {
+            let constr = algebraic_to_quadratic_symbolic_expression(&constr.expr);
+            let constr = try_remove_is_valid(&constr).unwrap_or(&constr);
+            zero_check_transformer.transform(constr.clone())
+        })
+        .collect_vec();
+
+    let constraints_by_variable = constraints
+        .iter()
+        .flat_map(|constr| {
+            constr
+                .referenced_unknown_variables()
+                .map(move |var| (var.clone(), constr))
+        })
+        .into_group_map();
+
+    // Collect the expressions that are used as addresses and try to solve
+    // a constraint about them in the constraint system.
+    let memory_addresses = memory_bus_interactions
+        .iter()
+        .map(|bus| algebraic_to_quadratic_symbolic_expression(&bus.addr))
+        .flat_map(|addr| {
+            // Go through the constraints related to this address
+            // and try to solve for the address.
+            let expr = addr
+                .referenced_unknown_variables()
+                .flat_map(|v| constraints_by_variable.get(v).map(|constrs| constrs.iter()))
+                .flatten()
+                .unique_by(|constr| constr as *const _)
```

Comment:
What does this do? Is it required?

Reply:
Can we make the comparison based on some field rather than the memory location?

Reply:
We could compare by the index in the vector, but then it would be better to change `constraints_by_variable` to only map to the index, which has the disadvantage that we only have an implicit connection between the index and what it means. I think it's ok to compare by pointer, it's essentially the same as a full comparison just that we can have false negatives.